### PR TITLE
GDScript: Fix inner classes and preloaded scripts as types

### LINKED ISSUE #47063: GDScript 2.0: Preloaded classes and inner clases cannot be used as type declarations
**Godot version:**
4.0.dev 8e293bac4e4c75179c0691f358fd8ef57a2bcf93

**OS/device including version:**
Windows 10.0.19041.804

**Issue description:**
I declared an inner class, which has a function returning a value of its own type or a supertype (sibling types are for some reason not in scope. Maybe related?)

Then, I declared a function in the containing .gd file which returns a value of the inner type. However, the two types are not compatible.

Script:
```
@tool
extends EditorPlugin
const external_type_class: GDScript = preload("res://addons/imported_type_test/external_type.gd")
class Foo:
	var bar: external_type_class = null
```

The error happens for all other declarations as well, such a variables within function scope.

The error also happens for declarations of inner classes, for example:
`var bar: external_type_class.InnerClass`

Error:
```
"external_type_class" is a constant but does not contain a type.
```

This is a regression from 3.2, where code is able to annotate as a preloaded type.

**Steps to reproduce:**
1. Open the example project.
2. Try to enable the "type_test_1" plugin.
3. It will give the above error at imported_type_test / plugin.gd

**Minimal reproduction project:**
[ImportedTypeDecl.zip](https://github.com/godotengine/godot/files/6149941/ImportedTypeDecl.zip)

### PR DESCRIPTION
**Pull request by vnen**
Fix #47063

### CODE REVIEW

File: modules/gdscript/gdscript_analyzer.cpp [hunk 1]
```diff
@@ -2121,6 +2141,9 @@ void GDScriptAnalyzer::reduce_call(GDScriptParser::CallNode *p_call, bool is_awa
 
 		if (is_self && parser->current_function != nullptr && parser->current_function->is_static && !is_static) {
 			push_error(vformat(R"*(Cannot call non-static function "%s()" from static function "%s()".)*", p_call->function_name, parser->current_function->identifier->name), p_call->callee);
+		} else if (!is_self && base_type.is_meta_type && !is_static) {
+			base_type.is_meta_type = false; // For `to_string()`.
+			push_error(vformat(R"*(Cannot call non-static function "%s()" on the class "%s" directly. Make an instance instead.".)*", p_call->function_name, base_type.to_string()), p_call->callee);
```
Comment (akien-mga) [inline modules/gdscript/gdscript_analyzer.cpp:2146]: I'm still getting confused by the raw string syntax, but isn't there an extraneous `."` at the end of the string here?

Comment (vnen) [reply modules/gdscript/gdscript_analyzer.cpp:2146]: Ah, indeed. I probably added it mechanically.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
The unit test seems to fail.

**Comment by vnen** [comment]
Added a couple of commits to fix bugs and pass the tests.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
There seems to be memory leaks detected by LSAN.

**Comment by vnen** [comment]
I fixed the leak but now something is being freed too soon and crashing. Not sure what's happening, seems like there's a deeper problem with the system.

**Comment by qarmin** [comment]
Looks that crash should be fixed by https://github.com/godotengine/godot/pull/51762

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**